### PR TITLE
[TensorCommon] Add get tensors cap common function @open sesame 11/18 16:59 

### PR DESCRIPTION
**Pull request by gichan-jang**
Add get tensors cap common function.
If the number of tensors is 1, it is to cover the case that only othenr/tensors is supported by peer pad.

Signed-off-by: gichan-jang <gichan2.jang@samsung.com>

Self evaluation:
1. Build test: [ *]Passed [ ]Failed [ ]Skipped
2. Run test: [ *]Passed [ ]Failed [ ]Skipped

### CODE REVIEW

File: gst/nnstreamer/include/nnstreamer_plugin_api.h [hunk 1]
```diff
@@ -351,6 +351,23 @@ gst_tensor_get_type (const gchar * typestr);
...
+extern GstCaps *
+gst_tensors_get_caps (const GstTensorsConfig * config, GstPad * pad);
+
+/**
+ * @brief Check whether pad's caps support other/tensor or not
+ * @param pad GstPad to check if it supports other/tensor
+ * @return TRUE if other/tensor, FALSE if not
+ */
+extern gboolean
+gst_tensor_peer_has_tensor_caps (GstPad * pad);
```
Comment (dongju-chae) [inline gst/nnstreamer/include/nnstreamer_plugin_api.h:369]: How about `gst_pad_peer_has_tensor_caps`?

Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api.h]: Changed to 'gst_pad_has_tensor_caps'.
Thamks

**Review by dongju-chae — APPROVED** [review]
LGTM

**Review by myungjoo — CHANGES_REQUESTED** [review]
Need test cases for ```gst_pad_has_tensor_caps ()``` and ```gst_tensors_get_caps ()``` on pads with:

1. "any"
2. "other/tensors"
3. "other/tensor"
4. "other/tensor, video, ..."
5. "other/tensor", "other/tensors", ...
6. "other/tensors,num_tensors=3, ..." (partially filled)
...

File: gst/nnstreamer/tensor_common.c [hunk 2]
```diff
@@ -857,6 +857,65 @@ gst_tensors_config_from_structure (GstTensorsConfig * config,
...
+ * @param pad GstPad to check if it supports other/tensor
+ * @param config tensors config structure
+ * @return caps for given config and pad
+ */
+GstCaps *
+gst_tensors_get_caps (GstPad * pad, const GstTensorsConfig * config)
+{
+  GstCaps *caps = NULL;
+
+  if (config->info.num_tensors == 1 && gst_pad_has_tensor_caps (pad)) {
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:871]: What is pad has "other/tensor,type=int32" and config is "other/tensor,type=int64"?

File: gst/nnstreamer/tensor_common.c [hunk 3]
```diff
@@ -857,6 +857,65 @@ gst_tensors_config_from_structure (GstTensorsConfig * config,
...
+    GstTensorConfig tensor_config;
+    tensor_config.info = config->info.info[0];
+    tensor_config.rate_n = config->rate_n;
+    tensor_config.rate_d = config->rate_d;
+    caps = gst_tensor_caps_from_config (&tensor_config);
+  } else {
+    caps = gst_tensors_caps_from_config (config);
```
Comment (myungjoo) [inline gst/nnstreamer/tensor_common.c:878]: Can you guarantee that this is ok if pad already has (partially) configured GSTCAP? or has some ranges in its values? (types or dims)

**Comment by gichan-jang** [comment]
> Need test cases for `gst_pad_has_tensor_caps ()` and `gst_tensors_get_caps ()` on pads with:
> 
> 1. "any"
> 2. "other/tensors"
> 3. "other/tensor"
> 4. "other/tensor, video, ..."
> 5. "other/tensor", "other/tensors", ...
> 6. "other/tensors,num_tensors=3, ..." (partially filled)
>    ...

I did not cover all cases, but I added test cases.

**Comment by myungjoo** [comment]
> > Need test cases for `gst_pad_has_tensor_caps ()` and `gst_tensors_get_caps ()` on pads with:
> > 
> > 1. "any"
> > 2. "other/tensors"
> > 3. "other/tensor"
> > 4. "other/tensor, video, ..."
> > 5. "other/tensor", "other/tensors", ...
> > 6. "other/tensors,num_tensors=3, ..." (partially filled)
> >    ...
> 
> I did not cover all cases, but I added test cases.

Please add unit tests that call ```gst_pad_has_tensor_caps ()``` and ```gst_pad_has_tensors_caps``` directly with GstPad objects.
Then, you can test with "any".

For better cap analysis, I'd suggest to use intersect methods; if (CAP-A intersect 'other/tensor') is not NULL, then CAP-A is compatible with 'other/tensor'.

**Comment by gichan-jang** [comment]
> > > Need test cases for `gst_pad_has_tensor_caps ()` and `gst_tensors_get_caps ()` on pads with:
> > > 
> > > 1. "any"
> > > 2. "other/tensors"
> > > 3. "other/tensor"
> > > 4. "other/tensor, video, ..."
> > > 5. "other/tensor", "other/tensors", ...
> > > 6. "other/tensors,num_tensors=3, ..." (partially filled)
> > >    ...
> > 
> > 
> > I did not cover all cases, but I added test cases.
> 
> Please add unit tests that call `gst_pad_has_tensor_caps ()` and `gst_pad_has_tensors_caps` directly with GstPad objects.
> Then, you can test with "any".
> 
> For better cap analysis, I'd suggest to use intersect methods; if (CAP-A intersect 'other/tensor') is not NULL, then CAP-A is compatible with 'other/tensor'.

Right.

**Comment by gichan-jang** [comment]
> Please add unit tests that call `gst_pad_has_tensor_caps ()` and `gst_pad_has_tensors_caps` directly with GstPad objects.
> Then, you can test with "any".

  It was difficult to use the GstPad object directly, so I used fakesrc and fakesink.

> For better cap analysis, I'd suggest to use intersect methods; if (CAP-A intersect 'other/tensor') is not NULL, then CAP-A is compatible with 'other/tensor'.

 Changed to use caps intersection.

File: gst/nnstreamer/include/nnstreamer_plugin_api.h [hunk 4]
```diff
@@ -351,6 +351,23 @@ gst_tensor_get_type (const gchar * typestr);
 extern const gchar *
 gst_tensor_get_type_string (tensor_type type);
 
+/**
+ * @brief Get tensors caps from tensors config and peer caps of the src pad
+ * @param config tensors config structure
+ * @param pad GstPad to check if it supports other/tensor
+ * @return caps for given config and pad
```
[the remark this answers is not in the document]
Comment (gichan-jang) [reply gst/nnstreamer/include/nnstreamer_plugin_api.h]: Updated.

**Review by myungjoo — APPROVED** [review]

**Pull request merged**